From aaf2c642fa900e2f6b4ffa9b8c902845d78bd951 Mon Sep 17 00:00:00 2001
From: Yimin Peng <quic_yiminp@quicinc.com>
Date: Tue, 8 Aug 2023 17:03:22 +0800
Subject: [PATCH 3/3] firmware: qcom_scm: qcom scm userspace interface

Implementation of userspace interface to invoke
qcom_scm calls from userspace through ioctl calls.

Change-Id: Id876aba8117834c23eaa794a4362c379aabc7067
Signed-off-by: Aravind Vijayakumar <quic_aprasann@quicinc.com>
Signed-off-by: Yimin Peng <quic_yiminp@quicinc.com>
---
 drivers/firmware/qcom_scm.c  | 82 +++++++++++++++++++++++++++++++++++-
 include/uapi/misc/scmioctl.h | 25 +++++++++++
 2 files changed, 106 insertions(+), 1 deletion(-)
 create mode 100644 include/uapi/misc/scmioctl.h

diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index b269463100ba..b564d6315e35 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -29,7 +29,9 @@
 #include <linux/interrupt.h>
 #include <linux/spinlock.h>
 #include <linux/ktime.h>
-
+#include <linux/miscdevice.h>
+#include <uapi/misc/scmioctl.h>
+ 
 #include "qcom_scm.h"
 #include "qtee_shmbridge_internal.h"
 
@@ -105,6 +107,8 @@ static const char * const qcom_scm_convention_names[] = {
 
 static struct qcom_scm *__scm;
 
+struct miscdevice *scmdev;
+
 static int qcom_scm_clk_enable(void)
 {
 	int ret;
@@ -346,6 +350,80 @@ static bool __qcom_scm_is_call_available(struct device *dev, u32 svc_id,
 
 	return ret ? false : !!res.result[0];
 }
+static int qcom_scm_open(struct inode *inode, struct file *filp)
+{
+	pr_info("qcom scm open called\n");
+	return 0;
+}
+
+static long scm_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
+{
+	struct scm_hand_shake scm_data;
+	void __user *ip = (void __user *)arg;
+	struct qcom_scm_desc desc;
+	int id, i, no_of_args;
+	struct qcom_scm_res res;
+
+	if (cmd != SCM_HAND_SHAKE_IOCTL)
+		return -EFAULT;
+
+	if (copy_from_user(&scm_data, ip, sizeof(struct scm_hand_shake)))
+		return -EFAULT;
+
+	no_of_args = scm_data.arginfo & 0x0F;
+
+	if (scm_data.svc == 0x01 || scm_data.svc == 0x02 || scm_data.svc == 0x05
+			|| scm_data.svc == 0x06 || scm_data.svc == 0x0c
+			|| scm_data.svc == 0x0f || scm_data.svc == 0x10
+			|| scm_data.svc == 0x11 || scm_data.svc == 0x13 || scm_data.svc == 0x15) {
+		desc.svc = scm_data.svc;
+		desc.cmd = scm_data.cmd;
+		desc.owner = ARM_SMCCC_OWNER_SIP;
+		desc.arginfo = scm_data.arginfo;
+
+		for (id = 0; id < no_of_args; id++)
+			desc.args[id] = scm_data.args_buffer[id];
+
+		scm_data.ret =  qcom_scm_call(__scm->dev, &desc, &res);
+		pr_info("scm ioctl - ret: %d\n",  scm_data.ret);
+
+		for (i = 0; i < MAX_QCOM_SCM_RETS; i++)
+			scm_data.qcom_scm_res[i] = res.result[i];
+
+		if (copy_to_user(ip, &scm_data, sizeof(struct scm_hand_shake)))
+			return -EFAULT;
+		return 0;
+	} else {
+		pr_err("Unsupported scm service: %d and command:%d  \n", scm_data.svc, scm_data.cmd);
+		return -EFAULT;
+	}
+}
+
+static const struct file_operations qcom_scm_fops = {
+	.open = qcom_scm_open,
+	.unlocked_ioctl = scm_ioctl,
+};
+
+static int qcom_scm_dev_init(struct platform_device *pdev)
+{
+	int err;
+	char *devname = "scmnode";
+	struct device *dev = &pdev->dev;
+
+	scmdev = devm_kzalloc(dev, sizeof(struct miscdevice), GFP_KERNEL);
+	if (!scmdev)
+		return -ENOMEM;
+
+	scmdev->minor = MISC_DYNAMIC_MINOR;
+	scmdev->name = devname;
+	scmdev->fops = &qcom_scm_fops;
+	err = misc_register(scmdev);
+	if (err) {
+		pr_err("scm misc device creation failure: %d \n", err);
+		return err;
+	}
+	return 0;
+}
 
 /**
  * qcom_scm_set_warm_boot_addr() - Set the warm boot address for cpus
@@ -3042,6 +3120,8 @@ static int qcom_scm_probe(struct platform_device *pdev)
 	scm->restart_nb.priority = 130;
 	register_restart_handler(&scm->restart_nb);
 
+	qcom_scm_dev_init(pdev);
+
 	/*
 	 * If requested enable "download mode", from this point on warmboot
 	 * will cause the boot stages to enter download mode, unless
diff --git a/include/uapi/misc/scmioctl.h b/include/uapi/misc/scmioctl.h
new file mode 100644
index 000000000000..f68d640dff55
--- /dev/null
+++ b/include/uapi/misc/scmioctl.h
@@ -0,0 +1,25 @@
+/* SPDX-License-Identifier: GPL-2.0-only WITH Linux-syscall-note
+ * * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
+ *  */
+#ifndef __SCMIOCTL_H__
+#define __SCMIOCTL_H__
+
+#include <linux/types.h>
+
+#define SCM_HAND_SHAKE_IOCTL       _IOWR('R', 10, struct scm_hand_shake)
+
+#define MAX_QCOM_SCM_RESULT 3
+
+#define MAX_QCOM_SCM_IN 10
+
+struct scm_hand_shake {
+	unsigned int svc;
+	unsigned int cmd;
+	unsigned int arginfo;
+	unsigned int args_buffer[MAX_QCOM_SCM_IN];
+	unsigned int ret;
+	unsigned int arg_type;
+	unsigned int qcom_scm_res[MAX_QCOM_SCM_RESULT];
+};
+
+#endif
-- 
2.17.1

